apisix/kubernetes.lua

```diff
@@ -0,0 +1,330 @@
+        headers = {
+            ["Host"] = _apiserver.host .. ":" .. _apiserver.port,
+            ["Authorization"] = "Bearer " .. _apiserver.token,
+            ["Accept"] = "application/json",
+            ["Connection"] = "keep-alive"
+        }
+    })
+
+    core.log.info("--raw=" .. informer.path .. "?" .. informer:list_query())
+
+    if not res then
+        return false, "RequestError", err or ""
+    end
+
+    if res.status ~= 200 then
+        return false, res.reason, res:read_body() or ""
+    end
+    local body, err = res:read_body()
+    if err then
+        return false, "ReadBodyError", err
+    end
+
+    local data, _ = core.json.decode(body)
+    if not data or data.kind ~= informer.list_kind then
+        return false, "UnexpectedBody", body
+    end
+
+    informer.version = data.metadata.resourceVersion
+
+    if informer.on_added ~= nil then
+        for _, item in ipairs(data.items or empty_table) do
+            informer:on_added(item, _constants.ListDrive)
+        end
+    end
+
+    informer.continue = data.metadata.continue
+    if informer.continue ~= nil and informer.continue ~= "" then
+        list(httpc, informer)
+    end
+
+    return true, "Success", ""
+end
+
+local function watch(httpc, informer)
+    local max_watch_times = 5
+    for _ = 0, max_watch_times do
+        local watch_seconds = 1800 + math.random(9, 999)
+        informer.overtime = watch_seconds
+        local http_seconds = watch_seconds + 120
+        httpc:set_timeouts(2000, 3000, http_seconds * 1000)
+
+        local res, err = httpc:request({
+            path = informer.path,
+            query = informer:watch_query(),
+            headers = {
+                ["Host"] = _apiserver.host .. ":" .. _apiserver.port,
+                ["Authorization"] = "Bearer " .. _apiserver.token,
+                ["Accept"] = "application/json",
+                ["Connection"] = "keep-alive"
+            }
+        })
+
+        core.log.info("--raw=" .. informer.path .. "?" .. informer:watch_query())
+
+        if err then
+            return false, "RequestError", err
+        end
+
+        if res.status ~= 200 then
+            return false, res.reason, res:read_body() or ""
+        end
+
+        local remainder_body = ""
+        local body
+        local reader = res.body_reader
+        local gmatch_iterator
+        local captures
+        local captured_size = 0
+
+        while true do
+            body, err = reader()
+            if err then
+                return false, "ReadBodyError", err
+            end
+
+            if not body then
+                break
+            end
+
+            if #remainder_body ~= 0 then
+                body = remainder_body .. body
+            end
+
+            gmatch_iterator, err = ngx.re.gmatch(body, "{\"type\":.*}\n", "jao")
+            if not gmatch_iterator then
+                return false, "GmatchError", err
+            end
+
+            captures, err = gmatch_iterator()
```

Comment:
A chunk read each time may contain multiple events. I feel that you can consider looping gmatch_iterator() to read all events.

Reply:
TKS review , But gmatch_iterator() is already in a loop.

Reply:
@crazyMonkey1995 :
TKS review , I will double check and test ,fix it.